server/src/main/java/org/elasticsearch/search/aggregations/metrics/GlobalOrdCardinalityAggregator.java

```diff
@@ -0,0 +1,203 @@
+            String name,
+            ValuesSource.Bytes.WithOrdinals valuesSource,
+            int precision,
+            int maxOrd,
+            SearchContext context,
+            Aggregator parent,
+            Map<String, Object> metadata) throws IOException {
+        super(name, context, parent, metadata);
+        this.valuesSource = valuesSource;
+        this.counts = null;
+        this.collector = new OrdinalsCollector(context.bigArrays(), maxOrd);
+        this.precision = precision;
+    }
+
+    @Override
+    public ScoreMode scoreMode() {
+        return valuesSource.needsScores() ? ScoreMode.COMPLETE : ScoreMode.COMPLETE_NO_SCORES;
+    }
+
+    @Override
+    public LeafBucketCollector getLeafCollector(LeafReaderContext ctx,
+            final LeafBucketCollector sub) throws IOException {
+        this.collector.set(valuesSource.globalOrdinalsValues(ctx));
+        return collector;
+    }
+
+    @Override
+    protected void doPostCollection() throws IOException {
+        counts = new HyperLogLogPlusPlusSparse(precision, collector.bigArrays, collector.visitedOrds.size());
+        collector.postCollect(counts);
+        collector.close();
+        collector = null;
+    }
+
+    @Override
+    public double metric(long owningBucketOrd) {
+        return counts == null ? 0 : counts.cardinality(owningBucketOrd);
+    }
+
+    @Override
+    public InternalAggregation buildAggregation(long owningBucketOrdinal) {
+        if (counts == null || owningBucketOrdinal >= counts.maxOrd() || counts.cardinality(owningBucketOrdinal) == 0) {
+            return buildEmptyAggregation();
+        }
+        // We need to build a copy because the returned Aggregation needs remain usable after
+        // this Aggregator (and its HLL++ counters) is released.
+        AbstractHyperLogLogPlusPlus copy = counts.clone(owningBucketOrdinal, BigArrays.NON_RECYCLING_INSTANCE);
+        return new InternalCardinality(name, copy, metadata());
+    }
+
+    @Override
+    public InternalAggregation buildEmptyAggregation() {
+        return new InternalCardinality(name, null, metadata());
+    }
+
+    @Override
+    protected void doClose() {
+        Releasables.close(counts, collector);
+    }
+
+    @Override
+    public void collectDebugInfo(BiConsumer<String, Object> add) {
+        // maybe need to add something for global?
+        super.collectDebugInfo(add);
+        add.accept("empty_collectors_used", 0);
+        add.accept("numeric_collectors_used", 0);
+        add.accept("ordinals_collectors_used", 1);
+        add.accept("ordinals_collectors_overhead_too_high", 0);
+        add.accept("string_hashing_collectors_used", 0);
+    }
+
+    private static class OrdinalsCollector extends LeafBucketCollector implements Releasable {
+
+        private final BigArrays bigArrays;
+        private SortedSetDocValues values;
+        private final int maxOrd;
+        private ObjectArray<BitArray> visitedOrds;
```

Comment:
I'm not generally a fan of `ObjectArray` just because it means we're allocating something per bucket. It makes sense here though. Probably worth talking about.

I wonder if it'd be worth having two storage mechanisms here - a list of longs and transitioning into a BitArray. BitArray would take up a bunch of space if all you are doing is recording a hand full of big longs. But maybe this isn't really worth our time. This is already fairly complex to have these two collection mechanisms.

Reply:
I don't think they would be tremendously big. In theory the biggest array would be at precisions 18, where BitArray can be up to ~11KB. In most cases it BitArray will be much smaller so not sure if it is worthy.

Reply:
Yeah. Almost certainly not worth it now.